### PR TITLE
trying to add the functionality to handle oil/water polymer

### PR DESCRIPTION
**Pull request by GitPaean**
It does not work yet, so it is WIP.

### CODE REVIEW

**Comment by GitPaean** [comment]
Thanks @totto82  for help with this PR.

File: examples/flow.cpp [hunk 1]
```diff
@@ -166,8 +167,13 @@ int main(int argc, char** argv)
         }
         // Polymer case
         else if ( phases.active( Opm::Phase::POLYMER ) ) {
-            Opm::flowEbosPolymerSetDeck(*deck, *eclipseState, *schedule, *summary_config);
-            return Opm::flowEbosPolymerMain(argc, argv);
+            if ( phases.size() == 3 ) {
```
Comment (GitPaean) [inline examples/flow.cpp]: this part is not for temporary, more careful checking should be made here later.

**Comment by GitPaean** [comment]
It does not work yet, please comment or suggest if you find something wrong. It is much appreciated.

**Comment by GitPaean** [comment]
The error message is something like, basically, the polymerConcentration index is 3, which should be 2. 
```
**********************************************************************
*                                                                    *
*                      This is flow 2018.04-pre                      *
*                                                                    *
* Flow is a simulator for fully implicit three-phase black-oil flow, *
*             including solvent and polymer capabilities.            *
*          For more information, see http://opm-project.org          *
*                                                                    *
**********************************************************************


===============Saturation Functions Diagnostics===============

System:  Oil-Water system.
Relative permeability input format: Saturation Family I.
Number of saturation regions: 1



================ Starting main simulation loop ===============

flow: /home/kaib/OPM-test/debug/opm-material/opm/material/densead/Evaluation3.hpp:100: Opm::DenseAd::Evaluation<ValueT, 3>::Evaluation(const RhsValueType&, int) [with RhsValueType = double; ValueT = double]: Assertion `0 <= varPos && varPos < size' failed.
Aborted ```

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/simulators/flow_ebos_oilwater_polymer.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace Ewoms {
+namespace Properties {
+NEW_TYPE_TAG(EclFlowOilWaterPolymerProblem, INHERITS_FROM(EclFlowProblem));
+SET_BOOL_PROP(EclFlowOilWaterPolymerProblem, EnablePolymer, true);
+//! The indices required by the model
+SET_TYPE_PROP(EclFlowOilWaterPolymerProblem, Indices,
+              Ewoms::BlackOilTwoPhaseIndices<GET_PROP_VALUE(TypeTag, EnableSolvent)?1:0,
+                GET_PROP_VALUE(TypeTag, EnablePolymer)?1:0,
+                /*PVOffset=*/0, /*disabledCompIdx=*/2>);
```
Comment (totto82) [inline opm/simulators/flow_ebos_oilwater_polymer.cpp]: Try to change 
Polymer  + 2p is never tested. I guess I found a bug. 
Try this patch.  
```
/home/thsa/workspace/opm/ewoms/ewoms/models/blackoil/blackoiltwophaseindices.hh
-   static const int polymerConcentrationIdx  = solventSaturationIdx + numPolymers;
+  static const int polymerConcentrationIdx  = solventSaturationIdx + numSolvents;

```

Comment (GitPaean) [reply opm/simulators/flow_ebos_oilwater_polymer.cpp]: Thanks. I am creating a fix in ewoms side.

Comment (GitPaean) [reply opm/simulators/flow_ebos_oilwater_polymer.cpp]: My suggestion was something like 
```
-    static const int solventSaturationIdx  = PVOffset + numPhases;
+    static const int solventSaturationIdx  = numSolvents > 0 ? PVOffset + numPhases : -10000;
 
     //! Index of the primary variable for the first polymer
-    static const int polymerConcentrationIdx  = solventSaturationIdx + numPolymers;
+    static const int polymerConcentrationIdx  = numPolymers > 0 ?
+                (numSolvents > 0? solventSaturationIdx + numPolymers : PVOffset + numPhases) : -10000;
```

Comment (GitPaean) [reply opm/simulators/flow_ebos_oilwater_polymer.cpp]: >
```
/home/thsa/workspace/opm/ewoms/ewoms/models/blackoil/blackoiltwophaseindices.hh
-   static const int polymerConcentrationIdx  = solventSaturationIdx + numPolymers;
+  static const int polymerConcentrationIdx  = solventSaturationIdx + numSolvents;
```

I will take your suggestion. It is simpler.

**Comment by GitPaean** [comment]
It depends on OPM/ewoms#291 now.  I will create a case in opm-data to make a running test out of it to make this functionality protected.

**Comment by GitPaean** [comment]
jenkins build this ewoms=291 please

**Comment by GitPaean** [comment]
jenkins build this ewoms=291 please

**Comment by GitPaean** [comment]
jenkins build this ewoms=291 opm-data=309 please

**Comment by GitPaean** [comment]
I removed the WIP status, and it is ready for review and merging when done reviewing. It depends on the OPM/ewoms#291 and the data in OPM/opm-data#309 is required for testing.  Thanks @totto82 for the help.

**Comment by totto82** [comment]
This will increase the compile time. I don't think we can add all the different variants to flow.cpp. Maybe we need a switch in the build system so we can restrict the build to the versions we want?

**Comment by GitPaean** [comment]
> This will increase the compile time. I don't think we can add all the different variants to flow.cpp. Maybe we need a switch in the build system so we can restrict the build to the versions we want?

Yes and it is also a very good suggestion. For most of the time, we only need `flow_ebos_blackoil.cpp` .

**Comment by GitPaean** [comment]
The motive for this PR is that many people work on polymer usually only use oil water + polymer, someone even said gas is a distraction. But in many fields, gas is typically there.  Both oil/water + polymer and blackoil + polymer have their own usage.

**Comment by atgeirr** [comment]
> This will increase the compile time. I don't think we can add all the different variants to flow.cpp. Maybe we need a switch in the build system so we can restrict the build to the versions we want?

I support adding such an option. I don't think we should put that burden on this PR however.

Added compile time may be noticeable for weaker systems, but for multi-core building with enough RAM (approx 2 GB per core) to compile at least 6 in parallel, it should not matter much. We may eventually reach a point where our current approach of instantiating *everything* depending on model (since all depend on the almighty type tag) may need replacing. For flow-legacy, the linear solver parts only depended on the block size for example, we could adopt that again, to cut down recompile times. (Then changes to linear solver details would not require recompile of the model/problem/simulator, and vice versa.)

**Comment by GitPaean** [comment]
jenkins build this ewoms=291 opm-data=309 please

**Comment by GitPaean** [comment]
``` Took 9 min 35 sec on master ```

The test took 9.5 minutes, my impression is that it takes half hour or more  than one hour sometimes. Something just changed?

File: opm/simulators/flow_ebos_oilwater_polymer.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+#include <dune/common/parallel/mpihelper.hh>
+#endif
+
+namespace Ewoms {
+namespace Properties {
+NEW_TYPE_TAG(EclFlowOilWaterPolymerProblem, INHERITS_FROM(EclFlowProblem));
+SET_BOOL_PROP(EclFlowOilWaterPolymerProblem, EnablePolymer, true);
+//! The indices required by the model
+SET_TYPE_PROP(EclFlowOilWaterPolymerProblem, Indices,
+              Ewoms::BlackOilTwoPhaseIndices<GET_PROP_VALUE(TypeTag, EnableSolvent)?1:0,
```
Comment (atgeirr) [inline opm/simulators/flow_ebos_oilwater_polymer.cpp]: Minor formatting issues: space around `?` and `:` in two places, indentation is slightly off.

File: opm/simulators/flow_ebos_oilwater_polymer.hpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+#define FLOW_EBOS_OILWATER_POLYMER_HPP
+
+#include <opm/parser/eclipse/Deck/Deck.hpp>
+#include <opm/parser/eclipse/EclipseState/EclipseState.hpp>
+#include <opm/parser/eclipse/EclipseState/Schedule/Schedule.hpp>
+#include <opm/parser/eclipse/EclipseState/SummaryConfig/SummaryConfig.hpp>
+
+
+namespace Opm {
+  void flowEbosOilWaterPolymerSetDeck(Deck &deck, EclipseState& eclState, Schedule& schedule, SummaryConfig& summary_config);
```
Comment (atgeirr) [inline opm/simulators/flow_ebos_oilwater_polymer.hpp]: Put `&` next to the type not the variable name.

File: tests/update_reference_data.sh [hunk 5]
```diff
@@ -107,6 +107,15 @@ for test_name in ${tests}; do
       EGRID INIT SMSPEC UNRST UNSMRY
   fi
 
+  if grep -q "polymer_oilwater" <<< $test_name
```
Comment (atgeirr) [inline tests/update_reference_data.sh]: I think you must also add `polymer_oilwater` to line 23.

**Comment by atgeirr** [comment]
Only minor quibbles, otherwise mergeable.

**Comment by GitPaean** [comment]
comments are  fixed.

**Comment by GitPaean** [comment]
jenkins build this ewoms=291 opm-data=309 please

**Comment by atgeirr** [comment]
Thanks, merging!

**Comment by andlaus** [comment]
I'm a bit too late to this party, but to me it seems that instead of making the build system even more complex, separate binaries for the more exotic simulators -- e.g., `flow_oilwater_polymer` in this case -- should be added back with the corresponding compile units not being part of `libopmsimulators`. The unified flow binary binary has always been intended to only capture the most common cases, at least as I originally envisioned it.

Long-term, the *meaning* of conservation equations could be assigned at runtime while their *number* stays specified at compile time. Such a mechanism should probably not be a top priority for at least another year, though.

**Pull request merged**